packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -2101,145 +2070,305 @@ function commitRootImpl(root, renderPriorityLevel) {
+    // so to satisfy that we also process the parent's "deletions" array (the deletion of siblings).
+    commitBeforeMutationEffectsDeletions(fiber.deletions);
+    const parent = fiber.return;
+    if (parent) {
+      commitBeforeMutationEffectsDeletions(parent.deletions);
     }
-    if ((effectTag & Passive) !== NoEffect) {
-      // If there are passive effects, schedule a callback to flush at
-      // the earliest opportunity.
-      if (!rootDoesHavePassiveEffects) {
-        rootDoesHavePassiveEffects = true;
-        scheduleCallback(NormalSchedulerPriority, () => {
-          flushPassiveEffects();
-          return null;
-        });
-      }
+
+    // Check to see if the focused element was inside of a hidden (Suspense) subtree.
+    // TODO: Move this out of the hot path using a dedicated effect tag.
+    if (
+      fiber.tag === SuspenseComponent &&
+      isSuspenseBoundaryBeingHidden(current, fiber) &&
+      doesFiberContain(fiber, focusedInstanceHandle)
+    ) {
+      shouldFireAfterActiveInstanceBlur = true;
+      beforeActiveInstanceBlur();
+    }
+  }
+
+  if ((effectTag & Snapshot) !== NoEffect) {
+    setCurrentDebugFiberInDEV(fiber);
+    commitBeforeMutationEffectOnFiber(current, fiber);
+    resetCurrentDebugFiberInDEV();
+  }
+
+  if ((effectTag & Passive) !== NoEffect) {
+    // If there are passive effects, schedule a callback to flush at
+    // the earliest opportunity.
+    if (!rootDoesHavePassiveEffects) {
+      rootDoesHavePassiveEffects = true;
+      scheduleCallback(NormalSchedulerPriority, () => {
+        flushPassiveEffects();
+        return null;
+      });
     }
-    nextEffect = nextEffect.nextEffect;
   }
 }
 
-function commitMutationEffects(root: FiberRoot, renderPriorityLevel) {
-  // TODO: Should probably move the bulk of this function to commitWork.
-  while (nextEffect !== null) {
-    setCurrentDebugFiberInDEV(nextEffect);
+function commitBeforeMutationEffectsDeletions(deletions: Array<Fiber>) {
+  for (let i = 0; i < deletions.length; i++) {
+    const fiber = deletions[i];
 
-    const effectTag = nextEffect.effectTag;
+    // TODO (effects) It would be nice to avoid calling doesFiberContain()
+    // Maybe we can repurpose one of the subtreeTag positions for this instead?
+    // Use it to store which part of the tree the focused instance is in?
+    // This assumes we can safely determine that instance during the "render" phase.
 
-    if (effectTag & ContentReset) {
-      commitResetTextContent(nextEffect);
+    if (doesFiberContain(fiber, ((focusedInstanceHandle: any): Fiber))) {
+      shouldFireAfterActiveInstanceBlur = true;
+      beforeActiveInstanceBlur();
     }
+  }
+}
 
-    if (effectTag & Ref) {
-      const current = nextEffect.alternate;
-      if (current !== null) {
-        commitDetachRef(current);
-      }
+function commitMutationEffects(
+  fiber: Fiber,
+  root: FiberRoot,
+  renderPriorityLevel,
+) {
```

Comment:
What if we called `commitMutationEffectsDeletions(fiber.deletions, ...)` here instead of in `commitMutationEffectsImpl`?

Reply:
We talked about this! I think the reason why we didn't proceed was because refactoring it was a bit annoying and Brian wanted to get your response to the approach before we committed to doing this?

Reply:
Having a "duh" moment now for not coming to this conclusion earlier.

Good suggestion. Thanks!